### PR TITLE
[DR-2678] TDR retrieveSnapshotPolicies should gracefully handle inaccessible workspaces

### PR DESCRIPTION
**Pull request by okotsopoulos**
https://broadworkbench.atlassian.net/browse/DR-2678

**Background**

Say a user has the needed permissions to view the policies on a snapshot.  When the snapshot has been export-by-copied to an inaccessible workspace, TDR’s `retrieveSnapshotPolicies` endpoint did not not gracefully handle that error, throwing on the first inaccessible workspace like so:

```
{
  "message": "404 Not Found: \"{\"causes\":[],\"message\":\"workspace ad0b2d69-e768-49b4-8954-b4cb55cff0cf does not exist or you do not have permission to use it\",\"source\":\"rawls\",\"stackTrace\":[],\"statusCode\":404}\"",
  "errorDetail": null
}
```

**Objective**

All workspaces to which a snapshot has been exported are represented in the `retrieveSnapshotPolicies` response, regardless of their accessibility.  If a workspace is inaccessible, we can't resolve its ID to a name and URL, but can still share the policy groups and supplement with the cause of inaccessibility.

This has implications for planned work to expand how destination workspaces are represented in TDR's Snapshot UI: https://broadworkbench.atlassian.net/browse/DR-2694

**Changes**

- Refactored `RawlsService.resolvePolicyEmails` to separate workspace policy responses based on accessibility
- Broke out main logic in `SnapshotsApiController.retrieveSnapshotPolicies` to a new method in `SnapshotService` of the same name, with a unit test
- Expanded `RawlsServiceTest.testRetrievePoliciesAndEmailsWorkspace` to offer more complete coverage -- now testing accessible workspaces, inaccessible workspaces, and non-workspace policies.

**Demo**

My [dev environment](https://jade-ok.datarepo-dev.broadinstitute.org/) is up-to-date with these changes.

I've added `DataRepoAdmins@dev.test.firecloud.org` and my Broad email as stewards for the following snapshot: https://jade-ok.datarepo-dev.broadinstitute.org/snapshots/786026c9-b850-49ec-818a-1e1bc4a7db50

As my test (admin) user, I successfully exported this snapshot to a new workspace: https://bvdp-saturn-dev.appspot.com/#workspaces/general-dev-billing-account/DR-2678-snapshot-export-to-inaccessible-workspaces/data

[Fetching policies](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/snapshots/retrieveSnapshotPolicies) as my test user -- workspace is accessible:
```
{
  "policies": [
    {
      "name": "admin",
      "members": [
        "DataRepoAdmins@dev.test.firecloud.org"
      ]
    },
    {
      "name": "reader",
      "members": [
        "policy-38ad2919-0b0c-4857-ba07-e38aadeb8169@dev.test.firecloud.org",
        "policy-c3f152ef-adf8-4c6a-86c4-723f114399af@dev.test.firecloud.org",
        "policy-a3fe2180-b011-4d6d-acba-cf4dae152f7e@dev.test.firecloud.org",
        "policy-e51a073f-a843-497d-8730-d0e33e7a1db0@dev.test.firecloud.org"
      ]
    },
    {
      "name": "discoverer",
      "members": []
    },
    {
      "name": "steward",
      "members": [
        "okotsopo@broadinstitute.org",
        "DataRepoAdmins@dev.test.firecloud.org"
      ]
    }
  ],
  "workspaces": [
    {
      "workspaceId": "81592eb3-fb87-4f8e-8536-af6ee92bd7eb",
      "workspaceName": "DR-2678-snapshot-export-to-inaccessible-workspaces",
      "workspaceNamespace": "general-dev-billing-account",
      "workspaceLink": "https://bvdp-saturn-dev.appspot.com/#workspaces/general-dev-billing-account/DR-2678-snapshot-export-to-inaccessible-workspaces",
      "workspacePolicies": [
        {
          "name": "writer",
          "members": [
            "policy-38ad2919-0b0c-4857-ba07-e38aadeb8169@dev.test.firecloud.org"
          ]
        },
        {
          "name": "project-owner",
          "members": [
            "policy-c3f152ef-adf8-4c6a-86c4-723f114399af@dev.test.firecloud.org"
          ]
        },
        {
          "name": "reader",
          "members": [
            "policy-a3fe2180-b011-4d6d-acba-cf4dae152f7e@dev.test.firecloud.org"
          ]
        },
        {
          "name": "owner",
          "members": [
            "policy-e51a073f-a843-497d-8730-d0e33e7a1db0@dev.test.firecloud.org"
          ]
        }
      ]
    }
  ],
  "inaccessibleWorkspaces": []
}
```

[Fetching policies](https://jade-ok.datarepo-dev.broadinstitute.org/swagger-ui.html#/snapshots/retrieveSnapshotPolicies) as my Broad user -- workspace is inaccessible:
```
{
  "policies": [
    {
      "name": "admin",
      "members": [
        "DataRepoAdmins@dev.test.firecloud.org"
      ]
    },
    {
      "name": "reader",
      "members": [
        "policy-38ad2919-0b0c-4857-ba07-e38aadeb8169@dev.test.firecloud.org",
        "policy-c3f152ef-adf8-4c6a-86c4-723f114399af@dev.test.firecloud.org",
        "policy-a3fe2180-b011-4d6d-acba-cf4dae152f7e@dev.test.firecloud.org",
        "policy-e51a073f-a843-497d-8730-d0e33e7a1db0@dev.test.firecloud.org"
      ]
    },
    {
      "name": "discoverer",
      "members": []
    },
    {
      "name": "steward",
      "members": [
        "okotsopo@broadinstitute.org",
        "DataRepoAdmins@dev.test.firecloud.org"
      ]
    }
  ],
  "workspaces": [],
  "inaccessibleWorkspaces": [
    {
      "workspaceId": "81592eb3-fb87-4f8e-8536-af6ee92bd7eb",
      "workspacePolicies": [
        {
          "name": "writer",
          "members": [
            "policy-38ad2919-0b0c-4857-ba07-e38aadeb8169@dev.test.firecloud.org"
          ]
        },
        {
          "name": "project-owner",
          "members": [
            "policy-c3f152ef-adf8-4c6a-86c4-723f114399af@dev.test.firecloud.org"
          ]
        },
        {
          "name": "reader",
          "members": [
            "policy-a3fe2180-b011-4d6d-acba-cf4dae152f7e@dev.test.firecloud.org"
          ]
        },
        {
          "name": "owner",
          "members": [
            "policy-e51a073f-a843-497d-8730-d0e33e7a1db0@dev.test.firecloud.org"
          ]
        }
      ],
      "error": {
        "message": "404 Not Found: \"{\"causes\":[],\"message\":\"workspace 81592eb3-fb87-4f8e-8536-af6ee92bd7eb does not exist or you do not have permission to use it\",\"source\":\"rawls\",\"stackTrace\":[],\"statusCode\":404}\"",
        "errorDetail": null
      }
    }
  ]
}
```

### CODE REVIEW

File: src/main/java/bio/terra/service/rawls/RawlsService.java [hunk 1]
```diff
@@ -26,21 +29,44 @@ public RawlsService(TerraConfiguration terraConfiguration, RawlsClient rawlsClie
...
+              List<ResourcePolicyModel> policies = entry.getValue();
+              try {
+                WorkspaceResponse workspace = rawlsClient.getWorkspace(workspaceId, userRequest);
+                accessible.add(
+                    workspace.toWorkspacePolicyModel(policies, terraConfiguration.getBasePath()));
+              } catch (Exception ex) {
+                inaccessible.add(toInaccessibleWorkspacePolicyModel(workspaceId, policies, ex));
+              }
+            });
+    return new WorkspacePolicyModels(accessible, inaccessible);
```
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/rawls/RawlsService.java:62]: I elected to fork the stream and group workspaces by their accessibility in the response.

File: src/test/java/bio/terra/service/rawls/RawlsServiceTest.java [hunk 2]
```diff
@@ -87,77 +94,105 @@ public void setUp() throws Exception {
     // Mock out the lower level client in individual as needed
     when(samResourceApi.getApiClient()).thenAnswer(a -> apiClient);
 
-    workspaceId = UUID.randomUUID();
-    workspaceName = "testWorkspace";
-    workspaceNamepace = "testNamespace";
-
-    when(rawlsClient.getWorkspace(any(), any()))
```
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/rawls/RawlsServiceTest.java:94]: With this broad mocking, we didn't register that we never called `rawlsClient.getWorkspace` for our `workspaceId`.

**Review by samanehsan — APPROVED** [review]
Lgtm!

File: src/main/java/bio/terra/service/rawls/RawlsService.java [hunk 3]
```diff
@@ -26,21 +29,44 @@ public RawlsService(TerraConfiguration terraConfiguration, RawlsClient rawlsClie
...
-              return workspace.toWorkspacePolicyModel(
-                  entry.getValue(), terraConfiguration.getBasePath());
-            })
-        .collect(Collectors.toList());
+              List<ResourcePolicyModel> policies = entry.getValue();
+              try {
+                WorkspaceResponse workspace = rawlsClient.getWorkspace(workspaceId, userRequest);
+                accessible.add(
+                    workspace.toWorkspacePolicyModel(policies, terraConfiguration.getBasePath()));
+              } catch (Exception ex) {
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/rawls/RawlsService.java:58]: Is it ok for this to be general or should we catch unauthorized exceptions only and let others bubble up?  I'm not sure how those are modeled with the Rawls client

File: src/main/java/bio/terra/service/snapshot/SnapshotService.java [hunk 4]
```diff
@@ -511,6 +519,31 @@ public UUID getFirstSourceDatasetIdFromSnapshotId(
         .orElseThrow(() -> new DatasetNotFoundException("Source dataset for snapshot not found"));
   }
 
+  /**
+   * @param snapshotId snapshot UUID
+   * @param userReq authenticated user
+   * @return SAM-derived snapshot policies attributed to the user, including workspace resolution
+   *     where applicable.
+   */
+  public PolicyResponse retrieveSnapshotPolicies(
```
Comment (nmalfroy) [inline src/main/java/bio/terra/service/snapshot/SnapshotService.java:528]: thanks for moving this into the service.  Makes more sense here than the controller

**Review by nmalfroy — APPROVED** [review]
lgtm!  Your call on if we should catch the more specific exception

File: src/main/java/bio/terra/service/rawls/RawlsService.java (hunk 3 above)
Comment (okotsopoulos) [reply src/main/java/bio/terra/service/rawls/RawlsService.java:58]: Thanks for bringing this up, I think we should catch all exceptions to ensure that any issue with workspace resolution does not bring down the policy fetch.  The Rawls client (which we wrote) catches and rethrows generic `Exception`s.

Comment (nmalfroy) [reply src/main/java/bio/terra/service/rawls/RawlsService.java:58]: ah right right.  Good point.  Nevermind then :)

**Pull request merged**